### ISSUE TITLE
java.io.InvalidClassException: org.apache.spark.rdd.NewHadoopRDD

### ISSUE THREAD

**Issue by ryan-williams** [andypetrella/spark-notebook#346]
I tried running a SN built against 1.5.0 with a 1.5.1 `spark.yarn.jar` and I see all tasks failing with 

```
java.io.InvalidClassException: org.apache.spark.rdd.NewHadoopRDD; local class incompatible: stream classdesc serialVersionUID = 6559201105987478974, local class serialVersionUID = -8749574334148360781
	at java.io.ObjectStreamClass.initNonProxy(ObjectStreamClass.java:617)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1622)
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1517)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1771)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1350)
	at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:1990)
	at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1915)
```

I assume this is due to 1.5.1 vs. 1.5.0 incompatibilities, but lmk if you think something else is going on.

**Comment by andypetrella** [comment]
Mmmh this can be that or a mismatch of the hadoop deps maybe(?)

However, the master support 1.5.1, the generator is not yet updated for it, but you can easily build it from a clone, wdyt?

**Comment by ryan-williams** [comment]
OK, I think the original exception above may have had to do with other deps I was pulling in depending on Spark 1.4.1 and something about NewHadoopRDD maybe having changed from then to 1.5.1.

However, I'm still confused about how/whether I can run against different versions of Spark from a given SN deployment.

I ran a "yarn-client" notebook just now on my  and passed a 1.5.1 jar as `spark.yarn.jar` but the web UI still indicated that I was running a 1.5.0 Spark; my SN where I saw this was built from master a few weeks ago, but I selected the 1.5.0 option from the build generator because that's the latest that was available. I haven't tried building from a source checkout because the generator is so good, you should just add 1.5.1 support there!

Basically I am trying to track down some weird behavior in Spark and confirm whether it happens in Spark 1.5.0 and/or 1.5.1, so I need some more clarity on whether SN is letting me effectively toggle the Spark version via `spark.yarn.jar` or whether a given SN deployment is in some ways stuck on the specific Spark version it is built for.

**Comment by andypetrella** [comment]
Indeed, a spark notebook is built against a version of spark, because it packages the libs in it.

However, I'm thinking in how to remove this for future versions. The problem is that for the distro I like to have the spark notebook self contained for rapid tests on spark, hence having the deps already there is a must.

I'm not beaten yet, I know there will be an elegant solution soon for that ^^

**Comment by ryan-williams** [comment]
OK I think this is definitely a problem affecting SN, including a SN I just built from HEAD, and I have a small repro for it below: any time a task uses a `NewHadoopRDD` I see this exception.

For example, I'll make a tiny file in HDFS:

```
echo "foo" > >(hadoop fs -put - bar)
```

From `spark-shell`, I have no problem loading it via [`wholeTextFiles`](https://github.com/apache/spark/blob/v1.5.1/core/src/main/scala/org/apache/spark/SparkContext.scala#L862), which uses the "new" Hadoop file API:

```
$ $SPARK_HOME/bin/spark-shell --master=yarn-client
scala> sc.wholeTextFiles("bar").count
…
res0: Long = 1
```

However, same thing in SN fails:

![](http://cl.ly/image/2W2a1k1M2l1u/Screen%20Shot%202015-10-29%20at%205.15.18%20PM.png)

[Full exception stack trace]().

Really interesting: note that the `stream classdesc serialVersionUID` and `local class serialVersionUID` I'm seeing today are **reversed** from what I was seeing when I originally reported this issue; something very fishy is going on here.

[This SO](http://stackoverflow.com/a/14504364/544236) paints an ominous picture:

> it is strongly recommended that all serializable classes explicitly declare serialVersionUID values, since the default serialVersionUID computation is highly sensitive to class details that may vary depending on compiler implementations, and can thus result in unexpected InvalidClassExceptions during deserialization.

Note that NewHadoopRDD does not declare a `serialVersionUID`, and SN may very well be using a different compiler (being built by SBT) than Spark (which is typically packaged with Maven, afaik).

What are your thoughts about this? Want to try building spark-notebook via Maven...?

**Comment by ryan-williams** [comment]
OK I think I found the issue: I was using a Spark 1.5.0 `spark.yarn.jar` though SN was built on Spark 1.5.1!

**Comment by andypetrella** [comment]
woooohooooo good catch man!!!

This versioning thingy is really a PITA...

**Comment by ryan-williams** [comment]
This is also consistent with the reversed `serialVersionUID`s from what I was originally seeing: when I first filed this issue I was trying to use a 1.5.1 `spark.yarn.jar` on a 1.5.0 SN!

**Comment by andypetrella** [comment]
exactly, actually I had the same with scala version... cassandra connector using scala 2.10.5.
I had clashes between serial for None and None. Hence telling me that None is not an Option oO.

yeaaaaaah

**Comment by TopSpoofer** [comment]
i got this error too， but my spark lib is 1.5.0， and my cluster was built by cdh-5.5.1(spark is 1.5.0).
so, i don't know what the problem！

**Comment by akshaya18j** [comment]
hey, I do face the same problem, any solution wld be vry grtful.! 
I'm new to spark and hadoop, I ve installed spark-2.0.2 and hadoop-2.6.0 in my system bt whr sparks's hadoop version is 2.4.0 wich is mentioned in its pom.xml.

When i execute the code in eclipse in my system it doesnt throw this error, bt when i use a client machine with all jars(spark-2.0 & sparks's hadoop-2.4
![imageedit_2_2408978208](https://cloud.githubusercontent.com/assets/17783040/21889109/4be34e7e-d8ed-11e6-970a-bd57de5004f5.jpg)
) copied and execute the same code it throws java.io.InvalidClassException.
